src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java

```diff
@@ -1,25 +1,134 @@
+        writeLock = readWriteLock.writeLock();
+        initProducts();
+    }
+
     @Override
     public Product getProduct(Long id) {
-        throw new RuntimeException("Not implemented");
+        readLock.lock();
+        try {
+            return getOptionalOfProduct(id)
+                    .orElseThrow(() -> new RuntimeException("No such element was found"));
+        } finally{
+            readLock.unlock();
+        }
     }
 
     @Override
     public List<Product> findProducts() {
-        throw new RuntimeException("Not implemented");
+        readLock.lock();
+        try {
+            return products.stream()
+                    .filter(this::productInStock)
+                    .filter(this::productPriceNotNull)
+                    .collect(Collectors.toList());
+        } finally {
+            readLock.unlock();
+        }
+
+    }
+
+    private boolean productPriceNotNull(Product product){
+        return product.getPrice()!=null;
+    }
+
+    private boolean productInStock(Product product){
+        return product.getStock()>0;
     }
 
     @Override
-    public void save(Product product) {
-        throw new RuntimeException("Not implemented");
+    public void save(Product product){
+        writeLock.lock();
+        try {
+            if(product.getId() == null) {
+                saveProduct(product);
+                return;
+            }
+
+            Optional<Product> optional = getOptionalOfProduct(product.getId());
+            if (optional.isEmpty()) {
+                saveProduct(product);
+            } else {
+                changeProduct(optional.get(), product);
+            }
+        }finally {
+            writeLock.unlock();
+        }
+    }
+
+    private void saveProduct(Product product){
+        product.setId(productId++);
+        products.add(product);
+    }
+
+    /**this method uses inner class and replaces old values of existed product with new*/
+    private void changeProduct(Product existed, Product product){
+        existed.changer()
+                .code(product.getCode())
+                .imageUrl(product.getImageUrl())
+                .currency(product.getCurrency())
+                .price(product.getPrice())
+                .description(product.getDescription())
+                .stock(product.getStock());
+    }
+
+    /**return optional of product by id for next checking of existence */
+    private Optional<Product> getOptionalOfProduct(Long id){
+        return products.stream()
+                .filter(product -> id.equals(product.getId()))
+                .findAny();
     }
 
     @Override
     public void delete(Long id) {
-        throw new RuntimeException("Not implemented");
+        writeLock.lock();
+        try {
+            if (getOptionalOfProduct(id).isEmpty()) {
+                throw new RuntimeException("No such element was found");
+            }
+            products.remove(getOptionalOfProduct(id).get());
```

Comment:
Create separate variable for getOptionalOfProduct(id) to prevent calling the method twice